Zend/zend_enum.c

```diff
@@ -0,0 +1,129 @@
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_compile.h"
+
+#define ZEND_ENUM_PROPERTY_ERROR() \
+	zend_throw_error(NULL, "Enum properties are immutable")
+
+#define ZEND_ENUM_DISALLOW_MACIG_METHOD(propertyName, methodName) \
+	do { \
+		if (ce->propertyName) { \
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include %s", methodName); \
+		} \
+	} while (0);
+
+static zend_object_handlers enum_handlers;
+
+zend_object *zend_enum_new(zval *result, zend_class_entry *ce, zval *case_name_zv, zval *scalar_zv)
+{
+	// Temporarily remove the ZEND_ACC_ENUM flag to allow instantiation
+	// FIXME: I'm sure we'll find a cleaner solution
+	ce->ce_flags &= ~ZEND_ACC_ENUM;
+	object_init_ex(result, ce);
+	ce->ce_flags |= ZEND_ACC_ENUM;
+
+	zend_object *zobj = Z_OBJ_P(result);
+	zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_CASE), case_name_zv, NULL);
+	if (scalar_zv != NULL) {
+		zend_string *value_property_name = zend_string_init("value", strlen("value"), 0);
+		zobj->handlers->write_property(zobj, value_property_name, scalar_zv, NULL);
+		zend_string_free(value_property_name);
+	}
+
+	zobj->handlers = &enum_handlers;
+
+	return zobj;
+}
+
+static void zend_verify_enum_properties(zend_class_entry *ce)
+{
+	zend_property_info *property_info;
+
+	ZEND_HASH_FOREACH_PTR(&ce->properties_info, property_info) {
+		if (zend_string_equals_literal(property_info->name, "case")) {
+			continue;
+		}
+		if (
+			ce->enum_primitive_type != IS_UNDEF
+			&& zend_string_equals_literal(property_info->name, "value")
+		) {
+			continue;
+		}
+		// FIXME: File/line number for traits?
+		zend_error_noreturn(E_COMPILE_ERROR, "Enums may not include member variables");
+	} ZEND_HASH_FOREACH_END();
+}
+
+static void zend_verify_enum_magic_methods(zend_class_entry *ce)
+{
+	// Only __get, __call and __invoke are allowed
+
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(constructor, "__construct");
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(destructor, "__destruct");
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(clone, "__clone");
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(__set, "__set");
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(__unset, "__unset");
+	ZEND_ENUM_DISALLOW_MACIG_METHOD(__isset, "__isset");
```

Comment:
What's the rationale for allowing `__get` but not `__isset`? Both are magic read-only operations on properties, and it would make it inconvenient to not be able to write code such as `$x = Suit::Hearts->$optionalMagicProperty ?? 'default';` which internally uses both magic methods

If this is deliberate, the reason could be documented in the RFC

Reply:
Honestly, IMO even `__get` is questionable. @Crell Should we reconsider removing `__get`?

Reply:
The thinking was that `__get` was "safe" from an immutability POV (at least as safe as any method), and it offered a way to have property-esque behavior if desired, especially if combined with `match`.  The same logic would apply to `__isset`, I think, which would imply including that.

That said, there's nothing that can be done with a virtual property that couldn't be done with a method instead, so it's not a mandatory feature.  I'd call them both nice-to-haves.

Reply:
@Crell Ok, you decide. I think allowing `__Isset` should be as easy as removing one line (and adding a test). Not being able to use `??` probably doesn't make a lot of sense. So I think we should allow both or neither.

Reply:
Let's go with both then, unless someone can make a case for one of them being dangerous.  I'll note that in the RFC.

Reply:
Done, isset is now also allowed.